### PR TITLE
Make output have same options as input. Remove "numinputs", "numoutputs" APIs which are not needed now.

### PR DESCRIPTION
**Pull request by linkerzhang**
As a follow-up of PR #321 , this PR includes,
1. Make output have the same options as input. "Single", "Optional" and "Variadic".
2. Node that has "optional" outputs will also need to use "" (empty string) as a placeholder.
3. Remove "NumInputs", "NumOutputs" APIs which are not needed now. The API "OpSchema& NumInputs(std::set<int> allowed_input_nums);" and "OpSchema& NumOutputs(std::set<int> allowed_output_nums);" are kept.

Please let me know your comments and thoughts. Thank you!

### CODE REVIEW

File: onnx/defs/schema.cc [hunk 1]
```diff
@@ -133,9 +113,30 @@ void OpSchema::Verify(const NodeProto& node) const {
...
+                node.name(),
+                ") has more outputs (",
+                node.output_size(),
+                ") than declared (",
+                outputs_.size(),
+                ") in op definition.");
+        }
+    }
+
+    if (node.output(out_idx).empty() && (Single == outputs_[out_idx].GetOption())) {
```
Comment (dzhulgakov) [inline onnx/defs/schema.cc]: so do we allow optional outputs now?

File: onnx/defs/schema.cc [hunk 2]
```diff
@@ -478,36 +438,48 @@ OpSchema& OpSchema::FillUsing(std::function<void(OpSchema&)> populator) {
...
+      ++max_input_;
+      break;
+    case OpSchema::Optional:
+      ++max_input_;
+      break;
+    case OpSchema::Variadic:
+      // Only last input formal parameter could be variadic.
+      ENFORCE((inputs_.size() - 1) == i);
+      max_input_ = std::numeric_limits<int>::max();
+    }
```
Comment (dzhulgakov) [inline onnx/defs/schema.cc]: nit: break

File: onnx/defs/schema.cc [hunk 3]
```diff
@@ -478,36 +438,48 @@ OpSchema& OpSchema::FillUsing(std::function<void(OpSchema&)> populator) {
...
-    ENFORCE(inputs_.size() == max_input_);
+
+  // Calculate min/max number of inputs.
+  for (int i = 0; i < (int)(inputs_.size()); ++i) {
+    switch (inputs_[i].GetOption()) {
+    case OpSchema::Single:
+      ++min_input_;
+      ++max_input_;
+      break;
+    case OpSchema::Optional:
```
Comment (dzhulgakov) [inline onnx/defs/schema.cc]: this is the case only if the input appears in the end. If (for some reason) we have optional inputs in the middle, I wouldn't count them to the total

File: docs/Changelog.md [hunk 4]
```diff
@@ -600,7 +600,7 @@ opset_import {
 <dd>Which axis to concat on</dd>
 </dl>
 
-#### Inputs (1 - &#8734;)
+#### Inputs (0 - &#8734;)
```
Comment (dzhulgakov) [inline docs/Changelog.md]: so we need some annotation to signify that this variadic argument actually needs >=1 input. Shall we have Variadic and VariadicOptional?

basically your docs shouldn't change (it's a sign that the spec didn't change)

Comment (linkerzhang) [reply docs/Changelog.md]: This is really a great question!!! In my current implementation, variadic argument needs >= 0 input.

I was thinking of using bitwise options. Say Single=1, Optional=2, Variadic=4 (Here variadic needs >=1 input). However, I realized that only Optional and Variadic could be merged. So I didn't pick up the design and change variadic needs >=0 input (which means VariadicOptional in your idea).

What do you think?
1) Keep current design to have Variadic needs >=0 input. or,
2) Make Variadic needs >=1 input and add VariadicOptional needs >=0 input.

File: onnx/defs/schema.cc (hunk 1 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: yes, we do. In our current operators, BatchNormalization is the only one case to have optional outputs.

File: onnx/defs/schema.cc (hunk 3 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: Not quite get your opinion here please. Are you saying that,
1. We should only allow optional inputs at the end please?
2. If there's optional input defined in the middle (there're Single inputs afterwards), we'll just not use it? in this case, the operator registration is invalid?
Thank you!

File: onnx/defs/schema.cc (hunk 2 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: Thank you for the nice catch!

File: docs/Changelog.md (hunk 4 above)
Comment (linkerzhang) [reply docs/Changelog.md]: btw, @dzhulgakov , I put the comment in schema.h to show the meaning of each option as.
  // Formal parameter options.

  enum FormalParameterOption {

    // The input formal parameter is single and not optional.

    // Number of this input is 1.

    Single = 0,

    // The input formal parameter is single and optional.

    // Number of this input is 0 or 1.

    Optional = 1,

    // The input formal parameter is variadic.

    // Number of this input is [0, n].

    Variadic = 2,

  };

I can put it in *.md documents also. Let's make a decision on the two solutions above firstly.

**Comment by gramalingam** [comment]
I am trying to understand various use-cases and scenarios: 
(1) the model decides to ignore one of the outputs, since it won't use it: here, it could use empty string; then, the operator implementation will not fill in that output
(2) the operator implementation may return one output conditionally: now, what do we do in the case where the model specifies an output variable X and the operator implementation does not return a value for this output? What will be the value bound to X in subsequent places where X is used? Or, does model evaluation stop with an error/exception in this case?
(3) The model specifies multiple outputs X, Y, and Z: so, I guess in this case it is the operator implementation's responsibility to fill in the right values for each of X, Y, and Z? What if it doesn't have enough values to fill in for X, Y, and Z? Also, does this complicate the shape-inference computation?

On a related note: is a tensor with zero elements a valid runtime value? (We could use such a tensor in some of these cases, but this will complicate shape-inference aspects potentially.)

File: onnx/defs/schema.cc (hunk 3 above)
Comment (linkerzhang) [reply onnx/defs/schema.cc]: I would keep the semantic of "optional" same for both input and output. This means, allowing "optional" input and output in the middle with empty string specified.

**Comment by gramalingam** [comment]
Thanks @linkerzhang for the explanation in the call. Just recording some key points here.

This proposal looks good to me. I think that a separate top-level clarification/documentation/discussion of the notion of erroneous models would be helpful (specifically, errors in the models not caught by the type-system), but that is separate from this PR.

My understanding is that case (1) in previous message is not allowed: if an output is single, the model should not use empty-string for that. (However, I think we can easily allow this semantically, since we can treat the empty-string as a wild-card, meaning a new name not used anywhere else. But this can be a future generalization, since the implementations need to be adapted to do this.)

**Comment by linkerzhang** [comment]
OK, now, I'm stepping back a little bit and am convinced to take the semantics that "variadic" asks for >=1 input. "VariadicOptional" is not added as we're not using it. It could be added whenever it's needed.
@gramalingam @dzhulgakov .

File: onnx/defs/schema.cc (hunk 3 above)
Comment (dzhulgakov) [reply onnx/defs/schema.cc]: It's fine to allow them in the middle, I'm just saying that reporting min_input in this case is invalid. Imagine you have

0: single
1: optional
2: single

We should either disallow this case or make min_input=max_input=3 for it. Current code min_input=2 and it doesn't make sense. Shall we disallow it for now?

**Comment by linkerzhang** [comment]
It's fine to allow them in the middle, I'm just saying that reporting min_input in this case is invalid. Imagine you have
0: single
1: optional
2: single
We should either disallow this case or make min_input=max_input=3 for it. Current code min_input=2 and it doesn't make sense. Shall we disallow it for now?

This is a GOOD point!!! @gramalingam and I also discussed on this case. 

I personally think we should keep the min semantics as it is. In this case, min_input_ is still 2. max_input_ is 3. We should have other verification logic to ensure that node's inputs should be fed with 2 real inputs and one "" (empty string) in between. Otherwise, the min_input_ semantic is a little bit confusing, we say min number of this operator is 3, but the 2nd one is optional. It's weird. Thoughts?

**Comment by linkerzhang** [comment]
I'm merging this PR and will for sure work on the last comment if strong preferences are there about my opinions. Thank you all!

**Comment by gramalingam** [comment]
To expand on what @linkerzhang said: one straightforward way to do the check is to modify "verify" to add an extra check: if the number of actual inputs is less than the computed min-inputs, then we should verify that all the missing number of inputs at the end are "optional".

This is mostly an implementation detail (of how the check is done), so it doesn't matter much how it is done. @linkerzhang, if we let min_input be 3 in the above example, it is not unreasonable ... because it captures our rules for when an empty-string must be used. If we count the arguments specified, including empty string, we need to have a minimum of 3 to successfully pass verification (and not 2). Further, it is more efficient: instead of doing the "straightforward" check above at every node, we find how many of the trailing parameters are optional, and decrement min_input by it, once, when we compute min_input.

**Pull request merged**